TS : history definition error

Hi,

I use React with TS (I use [react-scripts-ts](https://github.com/wmonk/create-react-app-typescript) and when I define the Router history, I've got this error :
```
Error in ./src/index.tsx
(32,15): error TS2322: Type 'SynchronizedHistory' is not assignable to type 'History | undefined'.
  Type 'SynchronizedHistory' is not assignable to type 'History'.
    Types of property 'listenBefore' are incompatible.
      Type '(hook: TransitionHook) => UnsubscribeCallback' is not assignable to type '(hook: TransitionHook) => () => void'.
        Types of parameters 'hook' and 'hook' are incompatible.
          Type 'History.History.TransitionHook' is not assignable to type 'MobxReactRouter.TransitionHook'.
            Types of parameters 'location' and 'location' are incompatible.
              Type 'MobxReactRouter.Location' is not assignable to type 'History.Location'.
                Property 'query' is optional in type 'Location' but required in type 'Location'.
```

It comes from this line : `<Router history={history} >`
I did exactly like the documentation : `const history = syncHistoryWithStore(browserHistory, routingStore);
`

Thank you for your work !

--- a/__test__/types-test.tsx
+++ b/__test__/types-test.tsx
@@ -1,13 +1,22 @@
 // This contains sample code which tests the typings. This code does not run, but it is type-checked
-import { RouterStore, syncHistoryWithStore } from '../';
-import { History, Location } from '../';
+import { Router, browserHistory } from 'react-router';
+import { RouterStore, SynchronizedHistory, Location, syncHistoryWithStore } from '../';
 
-const browserHistory: History = undefined;
 const routerStore: RouterStore = new RouterStore();
-const history = syncHistoryWithStore(browserHistory, routerStore);
+const history: SynchronizedHistory = syncHistoryWithStore(browserHistory, routerStore);
 
 {
-  // browserHistory.unsubscribe(); 
+  <Router history={history}>
+    {/* routes */}
+  </Router>
+}
+
+{
+  { <Router history={browserHistory} /> }
+  { <Router history={routerStore.history} /> }
+}
+
+{
   history.unsubscribe();
 
   const unsubscribeFromStore = history.listen(location => console.log(location.pathname));
@@ -17,7 +26,7 @@ const history = syncHistoryWithStore(browserHistory, routerStore);
 }
 
 {
-  const history: History = routerStore.history;
+  const history: SynchronizedHistory = routerStore.history;
   history.createHref('location');
   history.createHref({ pathname: 'path', hash: '#1234' });
 

--- a/__test__/types.spec.js
+++ b/__test__/types.spec.js
@@ -14,8 +14,8 @@ beforeEach(() => {
 
 describe('types', () => {
   it('can diagnose types without errors', () => {
-    const files = [path.join(__dirname, '/types-test.ts')];
-    const options = { noEmit: true };
+    const files = [path.join(__dirname, '/types-test')];
+    const options = { noEmit: true, jsx: 'react' };
     const program = ts.createProgram(files, options);
     const diagnosticErrors = ts.getPreEmitDiagnostics(program)
       .filter((diagnostic) => diagnostic.category === ts.DiagnosticCategory.Error)

--- a/package.json
+++ b/package.json
@@ -39,6 +39,7 @@
   },
   "homepage": "https://github.com/alisd23/mobx-react-router#readme",
   "devDependencies": {
+    "@types/react-router": "^2.0.41",
     "babel-eslint": "^6.1.2",
     "babel-jest": "^15.0.0",
     "babel-loader": "^6.2.5",

--- a/types.d.ts
+++ b/types.d.ts
@@ -1,3 +1,6 @@
+// Type definitions for mobx-react-router 3.1
+// Project: https://github.com/alisd23/mobx-react-router
+
 declare namespace MobxReactRouter {
 
   export type Action = 'POP' | 'PUSH' | 'REPLACE';
@@ -53,27 +56,29 @@ declare namespace MobxReactRouter {
     unregisterTransitionHook(hook: TransitionHook): void;
   }
 
-  export type Location = {
+  export interface Location {
     pathname: string;
     search: string;
     action: string;
     key: string;
     state: any;
     query: any;
-    hash?: string
+    hash?: string;
     basename?: string;
-  };
+  }
 
-  export type LocationDescriptor = {
+  export interface LocationDescriptor {
     pathname?: string;
     search?: string;
     action?: string;
     state?: any;
     hash?: string;
     key?: string;
-  };
+  }
 
-  export type SynchronizedHistory = History & { unsubscribe?: UnsubscribeCallback };
+  export interface SynchronizedHistory extends History {
+    unsubscribe?: UnsubscribeCallback;
+  }
 
   export class RouterStore {
     location: Location;

--- a/types.d.ts
+++ b/types.d.ts
@@ -34,6 +34,23 @@ declare namespace MobxReactRouter {
     createLocation(path: string, action?: Action, key?: string): Location;
     createLocation(location: LocationDescriptor): Location;
     getCurrentLocation: () => Location;
+
+    /** @deprecated */
+    setState(path: string): void;
+    /** @deprecated */
+    setState(location: LocationDescriptor): void;
+    /** @deprecated use push() instead */
+    pushState(path: string): void;
+    /** @deprecated use push() instead */
+    pushState(location: LocationDescriptor): void;
+    /** @deprecated use replace() instead */
+    replaceState(path: string): void;
+    /** @deprecated use replace() instead */
+    replaceState(location: LocationDescriptor): void;
+    /** @deprecated */
+    registerTransitionHook(hook: TransitionHook): UnsubscribeCallback;
+    /** @deprecated */
+    unregisterTransitionHook(hook: TransitionHook): void;
   }
 
   export type Location = {
@@ -41,9 +58,9 @@ declare namespace MobxReactRouter {
     search: string;
     action: string;
     key: string;
-    hash: string
     state: any;
-    query?: any;
+    query: any;
+    hash?: string
     basename?: string;
   };
 

